### ISSUE TITLE
RPC: gettransaction - implementation discussion

### ISSUE THREAD

**Issue by sondreb** [stratisproject/StratisBitcoinFullNode#2985]
I'm working on some improvements to the RPC APIs to make them more aligned with the developer specification for Bitcoin. One of those I have looked into, is the gettransaction. The current implementation is not correct, afaik.

Interactive console and RPC help output from the Bitcoin daemon. I do think this is 0.16, and there is 0.17 documentation out as well.

http://chainquery.com/bitcoin-api/gettransaction

https://bitcoincore.org/en/doc/0.17.0/rpc/wallet/gettransaction/

https://bitcoin.org/en/developer-reference#gettransaction

From the last one, it says: "An object describing how the transaction affects the wallet".

And on the Amount field: "A positive number of bitcoins if this transaction increased the total wallet balance; a negative number of bitcoins if this transaction decreased the total wallet balance, or 0 if the transaction had no net effect on wallet balance"

This is my suggestion for new XML comment on the function:

**RPC method to return the effect an transaction has on the wallet. It doesn't show the raw Bitcoin transaction itself. Use getrawtransaction for that.**

And then to implement this, things get a bit funky, which is why I wanted to create an issue before I make a PR. I have not optimized the code yet, I just finally finalized this after many hours of work trying to figure out the APIs and how to make it correct.

First I implemented a new method on the WalletManager, which returns AccountHistory based on the transaction ID: **GetHistoryById(uint256 trxid, string walletName, string accountName = null)**.

![image](https://user-images.githubusercontent.com/309938/50242845-c8958f80-03cb-11e9-8d3c-5935dacfdbc5.png)

From there I flatten out some of the collections, and check if it is a change address or not.

If it is a change address, I set the "category" on details to "send", otherwise it is "receive". This variable is referenced later.

![image](https://user-images.githubusercontent.com/309938/50242964-114d4880-03cc-11e9-924f-73c923d86971.png)

And this is where it gets a tiny bit wonky, we have to query for Payments on the previous transaction, to get the proper values needed for the details array:

![image](https://user-images.githubusercontent.com/309938/50243283-165ec780-03cd-11e9-85a8-256b0b9a1a3b.png)

For the non-change address, we can simply grab the transaction amount:

![image](https://user-images.githubusercontent.com/309938/50243325-31313c00-03cd-11e9-9f6e-06f2442833fb.png)

And then finally return the model:

![image](https://user-images.githubusercontent.com/309938/50243342-3f7f5800-03cd-11e9-8c46-50e12597aa60.png)

While this does not implement all features, it adds "Confirmations" which was missing from before, and it lists a correct Details output, and correct Amount output.

I will be testing some more, with multiple transactions to same address, etc.

Please give some feedback, or else I'll just make a PR at some time based on what I have presented here.

**Comment by sondreb** [comment]
IsCoinStake should result in what value for category? The RPC help says only "send" and "receive", but bitcoin dot org says "generate", "immature" and "orphan" as alternatives. I don't think coinbase reward is "generate", but might be? I'm setting it to "stake" for now. Or perhaps "reward"?

**Comment by sondreb** [comment]
Another addition, support for stake transactions. Since the Amount is useless for these transactions, it is hard-coded to return the consensus value:

![image](https://user-images.githubusercontent.com/309938/50245691-6b9dd780-03d3-11e9-9dc4-f5a875c7b2ea.png)

**Comment by zeptin** [comment]
@sondreb Regarding the coinstake amount - don't forget the possible presence of transaction fees.

This seems to be how stratisX chooses the category, i.e. `generated`:

```
void WalletTxToJSON(const CWalletTx& wtx, Object& entry)
{
    int confirms = wtx.GetDepthInMainChain();
    entry.push_back(Pair("confirmations", confirms));
    if (wtx.IsCoinBase() || wtx.IsCoinStake())
        entry.push_back(Pair("generated", true));
    if (confirms > 0)
    {
        entry.push_back(Pair("blockhash", wtx.hashBlock.GetHex()));
        entry.push_back(Pair("blockindex", wtx.nIndex));
        entry.push_back(Pair("blocktime", (int64_t)(mapBlockIndex[wtx.hashBlock]->nTime)));
    }
    entry.push_back(Pair("txid", wtx.GetHash().GetHex()));
    entry.push_back(Pair("time", (int64_t)wtx.GetTxTime()));
    entry.push_back(Pair("timereceived", (int64_t)wtx.nTimeReceived));
    BOOST_FOREACH(const PAIRTYPE(string,string)& item, wtx.mapValue)
        entry.push_back(Pair(item.first, item.second));
}
```

I'm looking at the rest of what you've suggested.

**Comment by sondreb** [comment]
Thanks, I did this:

if (transaction.IsCoinStake.GetValueOrDefault(false) || transaction.IsCoinBase.GetValueOrDefault(false))

And then set it to "generated", as oppose to "stake". I added the BlockIndex also.

Compared to the stratisX code, the following are now missing in my implementation thus far:

"blocktime", "timereceived".

**Comment by mikedennis** [comment]
Yes looks like you are on the right track. For comparison this is what bitcoin core is doing (which is similar to stratisx): 
https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpcwallet.cpp#L95

```
static void WalletTxToJSON(interfaces::Chain& chain, interfaces::Chain::Lock& locked_chain, const CWalletTx& wtx, UniValue& entry)
{
    int confirms = wtx.GetDepthInMainChain(locked_chain);
    entry.pushKV("confirmations", confirms);
    if (wtx.IsCoinBase())
        entry.pushKV("generated", true);
    if (confirms > 0)
    {
        entry.pushKV("blockhash", wtx.hashBlock.GetHex());
        entry.pushKV("blockindex", wtx.nIndex);
        entry.pushKV("blocktime", LookupBlockIndex(wtx.hashBlock)->GetBlockTime());
    } else {
        entry.pushKV("trusted", wtx.IsTrusted(locked_chain));
    }
    uint256 hash = wtx.GetHash();
    entry.pushKV("txid", hash.GetHex());
    UniValue conflicts(UniValue::VARR);
    for (const uint256& conflict : wtx.GetConflicts())
        conflicts.push_back(conflict.GetHex());
    entry.pushKV("walletconflicts", conflicts);
    entry.pushKV("time", wtx.GetTxTime());
    entry.pushKV("timereceived", (int64_t)wtx.nTimeReceived);

    // Add opt-in RBF status
    std::string rbfStatus = "no";
    if (confirms <= 0) {
        LOCK(mempool.cs);
        RBFTransactionState rbfState = IsRBFOptIn(*wtx.tx, mempool);
        if (rbfState == RBFTransactionState::UNKNOWN)
            rbfStatus = "unknown";
        else if (rbfState == RBFTransactionState::REPLACEABLE_BIP125)
            rbfStatus = "yes";
    }
    entry.pushKV("bip125-replaceable", rbfStatus);

    for (const std::pair<const std::string, std::string>& item : wtx.mapValue)
        entry.pushKV(item.first, item.second);
}
```

**Comment by maciejzaleski** [comment]
Issue will be resolved as part of #2758
